### ISSUE TITLE
Elisions/replacements: complete discussion of accessibility

### ISSUE THREAD

**Issue by rebeccacremona** [harvard-lil/h2o#655]
Looks like we are missing some of the accessibility features we had before for elisions and replacements. See below for a detailed discussion. Errs on the side of too-detailed, for clarity. Not because I don't think people know what tabindexes are for 😉

**Comment by rebeccacremona** [comment]
Okay, here is an example of how an elision works for students on prod right now, with accessibility features highlighted. https://opencasebook.org/casebooks/25965-contracts/resources/1.1-john-d-r-leonard-plaintiff-v-pepsico-inc-defendant

Elision in p 34, hidden by default:
`<span role="button" tabindex="0" class="annotate elide" data-annotation-id="540594" aria-label="elided text" aria-expanded="false"></span>`

First, discussion about the html element:
- This element is conceptually a button, because it's an interactive element that causes an in-page action. Native html `<button>` elements come with most everything we need. I found that when I used an actual `<button>` element, the export code broke: the elision did not appear in exports at all.... the ellipses went missing, and the tests didn't fail!!! See https://github.com/harvard-lil/h2o/pull/422 and https://github.com/harvard-lil/h2o/issues/420. I wasn't up to altering the xlst to produce valid docx.... So instead, I retrofitted a `span` to behave just like a native `button`. On stage, I see that for elisions, a native button is now used!! And exporting w/ annotations seems to include the ellipses! Amazing!!! But I bring it up because *replacement* elisions, which are not now screen reader or keyboard accessible on stage, should behave just like these, and are currently using a `span` on stage. If Casey or Greg have not done some magic to fix the export code so that all buttons reliably appear, there is a risk that switching replacements from spans to buttons will cause them to soundlessly disappear from exports again.

- the `role="button"` is part of retrofitting the `span` and informs screen readers that this is a button, so it should announce it as such, and therefore informs screen reader users that they should expect it to behave like a button: it is interactive, it can be pressed using spacebar, and by pressing enter. However, `role=button` doesn't, by itself, achieve any of that: it's up to js to make it happen. Relevant for replacements, if they can't presently be switched from spans to buttons.

- the `tabindex=0` makes the `span` keyboard focusable like interactive elements (links, buttons, form elements....) should be. `0` means to put it in the tab order in the logical DOM order. You can put other integers there, but you shouldn't. Except "-1" sometimes. But we can get to that at some later point..... Without `tabindex=0`, nobody using a keyboard, or a switch, or etc. can navigate to and press the button. Relevant for replacements, if they can't presently be switched from spans to buttons. This is why the current replacement spans can't receive focus.

The  the rest:

- `aria-label="elided text"` since these "buttons" show a css-supplied ellipsis to sighted users, which is insufficiently descriptive for non-visual users, we provide an override for the button's label. Relevant for replacements: we will NOT need to provide an explicit label for the toggle button, while collapsed or expanded: it should probably be the replacement text while collapsed, and the word "hide" while expanded, as it is now on prod, both of which should be achieved by the browser without intervention from us. So, the aria-label that appears right now on stage for the button to "hide" an expanded replacement should probably not be there.

- `aria-expanded="false"`. So, this isn't a plain button..... this is a popup button, a button that when you press it, makes content nearby it appear, that was once hidden. But right now, that content is hidden, not displayed. The "aria-expanded" attribute communicates all that to a screen reader user. When they press it, we supply js that changes `aria-expanded="true"`. Relevant for both elisions and replacements: on stage, `aria-expanded="false"` is not present when hidden (though `aria-expanded="true"` IS present when expanded. 

So, our sample elision sounds like this in VoiceOver, when you arrive on it in prod: "elided text, button, collapsed"

Then, when you press it, you hear: "elided text, button, expanded". 

And then you can read along in the normal way.... and then, at the end, I put a screen-reader only `<span>` reading "end of elided text", right where sighted users stop seeing the special background color and font color indicating the elision is over:
`<span class="annotate elided revealed sr-only" data-annotation-id="540594">(end of elided text)</span>` The only important thing there is the class `sr-only`, which is a bootstrap class used for visually hiding text, but exposing it to screen readers, designed for just these kinds of hairy situations. Everything else is cruft related to the way annotations were implemented. 

The replacements sounds similar, except the button label, and except that it says "end of replaced text"  at the end.

I'm not completely satisfied with the UX of my solution, especially for long elisions like this example, or for communicating what a replacement is, but I think it's close enough for a first pass.

I'd recommend we try to add something at least as functional as this back in.

If useful, it was added initially in https://github.com/harvard-lil/h2o/pull/363/, and tweaked later.

**Comment by rebeccacremona** [comment]
actually, I note that while logged out, I can't expand replacement annotations even using my mouse.

**Comment by rebeccacremona** [comment]
- Replacement annotations should be buttons (either real `button` or retrofitted spans) for logged-out users
- Both elision buttons and replacement buttons need `aria-expanded="false"` when they are collapsed; ensure that js continues to toggle them to `aria-expanded="true"` when the buttons are pressed and the original text is revealed
- When elisions and replacement annotations are expanded, you have to mark the end of the selected text for screen reader users, who can't see the change in background color, as described above
- The `aria-label` on the "hide" button for replacement annotations is incorrect and should be removed; "hide" is fine for now, pending more thinking on replacement annotation UX
